### ISSUE TITLE
Migration Error - "Integrity constraint violation"

### ISSUE THREAD

**Issue by tigerchick** [doublesecretagency/craft-googlemaps#124]
When upgrading to 4.6.0 from 4.5.2 I get a migration error which causes the update to fail and for the site to go down:

_Integrity constraint violation: SQLSTATE[23000]: Integrity constraint violation...._

The only way to clear this is to revert that database and the composer.json / composer.lock files.

Sorry, I didn't take a full record of the error.. I did note however that the issue above occurred very many times in the stated error (presumably for each sql record that was migrated (or not!)

This on Craft 4.13.5

**Comment by lindseydiloreto** [comment]
Thanks for reporting! You are now the second person to report this issue, but the first person didn't have a pre-migration database backup.

I might have a solution for this, will post back here shortly.

**Comment by tigerchick** [comment]
Hi Lindsey.

Yep sure, thanks for looking into this so fast.
DB export can be downloaded here: 
https://icedrive.net/s/ZNiW5XbPjA84iZFf8RRFiafPa22b  (it’s about 16mb, 
so a little large for email).
Composer json files attached too.

Let me know if there’s anything else I can do to help,
Michael

Michael Rummey - Tigerchick Signature
Michael Rummey
Tigerchick Logo
t: 01968 675565
www.tigerchick.com <https://www.tigerchick.com>
Consider the environment before printing this email.This email (and any 
attachments) is confidential and may be privileged. It may be read, 
copied, and used only by the addressee. If you have received it in error 
please contact Tigerchick immediately. No responsibility or liability 
whatsoever will be accepted by Tigerchick for this email message and/or 
if the recipient's computer system is corrupted by information 
transmitted from the Tigerchick computer and/or for any delays or other 
losses arising out of or caused by transmissions.


------ Original Message ------
From "Lindsey DiLoreto" ***@***.***>
To "doublesecretagency/craft-googlemaps" 
***@***.***>
Cc "tigerchick" ***@***.***>; "Author" 
***@***.***>
Date 08/12/2024 16:26:55
Subject Re: [doublesecretagency/craft-googlemaps] Migration Error on 
updating from 4.5.2 to 4.6.0 (Issue #124)

**Comment by lindseydiloreto** [comment]
I just pushed a patch which (hopefully) resolves this issue.

When you get a chance, please grab the latest dev release and try again...

```js
"doublesecretagency/craft-googlemaps": "dev-v4-dev"
```

Please let me know if that does the trick for you! 🤞

**Comment by tigerchick** [comment]
Sorry for the delay. I just tried this but it failed again - with a different error this time:

```
One of Google Maps’s migrations failed.

Database Exception: SQLSTATE[HY000]: General error: 1832 Cannot change column 'siteId': used in a foreign key constraint 'fk_ajypmcpbnbdzmtunrmumevunblpjpljaerbl'
The SQL being executed was: ALTER TABLE googlemaps_addresses CHANGE siteId siteId int(11) NOT NULL
```

I'll email you the full output.

Edit: This "new" error actually seems to be the same error that was reported in the thread #125

**Comment by lindseydiloreto** [comment]
Yeah, looks like you're now hitting that other error. I suspect that both bugs were in there all along, and fixing yours just made the other one more visible.

It almost sounds like something didn't get rolled back properly. I'll probably need a full database export to get it resolved.

Will ping you via email to discuss further. 👍

**Comment by lindseydiloreto** [comment]
Actually, I _may_ have just been able to fix it blind. Grab the latest `dev` release and give it another shot...

```js
"doublesecretagency/craft-googlemaps": "dev-v4-dev"
```

Make sure you've rolled back the database before trying the migration again.

Let me know how it goes! 🤞

**Comment by tigerchick** [comment]
> Actually, I _may_ have just been able to fix it blind. Grab the latest `dev` release and give it another shot...
> 
> ```js
> "doublesecretagency/craft-googlemaps": "dev-v4-dev"
> ```
> 
> Make sure you've rolled back the database before trying the migration again.
> 
> Let me know how it goes! 🤞

Before I run this, can I confirm that is the correct dev version - v4 is the same number as before. (Deliberate, I'm sure, just want to double check!)

**Comment by lindseydiloreto** [comment]
Yes, that is correct. `v4-dev` is the branch where these changes are, and the `dev-` prefix allows you to get changes which haven't been tagged yet (like this one).

**Comment by tigerchick** [comment]
I can confirm that this now works correctly. Successful migration! Well done.

**Comment by lindseydiloreto** [comment]
Brilliant, thanks for confirming! 🍺

**Comment by lindseydiloreto** [comment]
Great news, this patch has been officially released in [v4.6.1](https://github.com/doublesecretagency/craft-googlemaps/releases/tag/4.6.1). 🎉

If you've experienced a failed migration, be sure to roll the database back before trying again.

Hope that helps, thanks again for reporting!
